### ISSUE TITLE
Error on provisioning - "message":"must not exist"

### ISSUE THREAD

**Issue by drasko** [telefonicaid/iotagent-mqtt#87]
Hi,
I am trying to provision device with:
```
(curl localhost:4041/iot/devices -s -S --header 'Content-Type: application/json' \
    --header 'Accept: application/json' --header 'Fiware-Service: mainflux' --header 'Fiware-ServicePath: /mqtt' -d @- | python -mjson.tool) <<EOF
{
    "devices": [ 
        { 
            "device_id": "DevID1", 
            "entity_name": "TheDevice1", 
            "entity_type": "DeviceType", 
            "attributes": [ 
                  { "id": "t", "name": "temperature", "type": "float" },
                  { "id": "h", "name": "humidity", "type": "float" }
            ],
            "lazy":[
                  { "id": "l", "name": "luminosity", "type": "percentage" }
            ],
            "commands": [
                  { "id": "t", "name": "turn", "type": "string" }
            ],
            "static_attributes": [
                  { "name": "serialID", "type": "02598347" }
            ]
        }
    ]
} 
EOF
```

And I am getting the error:
```
{
    "message": "Wrong syntax in request: Errors found validating request.",
    "name": "WRONG_SYNTAX"
}
```

Here is IOTA log:
```
drasko@Lenin:~/fiware/iotagent-mqtt$ node bin/iotagentMqtt.js 
time=2016-02-04T23:26:13.513Z | lvl=INFO | corr=n/a | trans=n/a | op=IoTAgentNGSI.Global | msg=Activating IOT Agent NGSI Library.
time=2016-02-04T23:26:13.515Z | lvl=INFO | corr=n/a | trans=n/a | op=IoTAgentNGSI.Global | msg=Falling back to Transient Memory registry for NGSI Library
time=2016-02-04T23:26:13.519Z | lvl=INFO | corr=n/a | trans=n/a | op=IoTAgentNGSI.NorthboundServer | msg=Starting IoT Agent listening on port [4041]
time=2016-02-04T23:26:13.520Z | lvl=DEBUG | corr=n/a | trans=n/a | op=IoTAgentNGSI.NorthboundServer | msg=Using config:

{
    "logLevel": "DEBUG",
    "contextBroker": {
        "host": "localhost",
        "port": "1026"
    },
    "server": {
        "port": 4041
    },
    "deviceRegistry": {
        "type": "memory"
    },
    "types": {},
    "service": "mainflux",
    "subservice": "/mqtt",
    "providerUrl": "http://localhost:4041",
    "deviceRegistrationDuration": "P1M",
    "defaultType": "Thing",
    "iotaVersion": "0.1.5"
}

time=2016-02-04T23:26:13.533Z | lvl=INFO | corr=n/a | trans=n/a | op=IoTAgentNGSI.ContextServer | msg=Loading NGSI Contect server routes
time=2016-02-04T23:26:13.534Z | lvl=DEBUG | corr=n/a | trans=n/a | op=n/a | msg=Loading access validation Templates
time=2016-02-04T23:26:13.541Z | lvl=INFO | corr=n/a | trans=n/a | op=IoTAgentMQTT.Agent | msg=IoT Agent services activated
time=2016-02-04T23:26:13.556Z | lvl=INFO | corr=n/a | trans=n/a | op=IoTAgentMQTT.Agent | msg=MQTT Client connected
time=2016-02-04T23:26:13.556Z | lvl=DEBUG | corr=n/a | trans=n/a | op=IoTAgentMQTT.Agent | msg=Recreating subscriptions for all devices
time=2016-02-04T23:26:13.557Z | lvl=INFO | corr=n/a | trans=n/a | op=IOTAMQTT.Executable | msg=MQTT IoT Agent started
time=2016-02-04T23:26:41.729Z | lvl=DEBUG | corr=n/a | trans=n/a | op=IoTAgentNGSI.NorthboundServer | msg=Request for path [/iot/devices] from [localhost:4041]
time=2016-02-04T23:26:41.730Z | lvl=DEBUG | corr=n/a | trans=n/a | op=IoTAgentNGSI.NorthboundServer | msg=Body:

{
    "devices": [
        {
            "device_id": "DevID1",
            "entity_name": "TheDevice1",
            "entity_type": "DeviceType",
            "attributes": [
                {
                    "id": "t",
                    "name": "temperature",
                    "type": "float"
                },
                {
                    "id": "h",
                    "name": "humidity",
                    "type": "float"
                }
            ],
            "lazy": [
                {
                    "id": "l",
                    "name": "luminosity",
                    "type": "percentage"
                }
            ],
            "commands": [
                {
                    "id": "t",
                    "name": "turn",
                    "type": "string"
                }
            ],
            "static_attributes": [
                {
                    "name": "serialID",
                    "type": "02598347"
                }
            ]
        }
    ]
}


time=2016-02-04T23:26:41.733Z | lvl=DEBUG | corr=n/a | trans=n/a | op=IoTAgentNGSI.RestUtils | msg=Errors found validating request: {"valid":false,"errors":[{"attribute":"additionalProperties","property":"devices.0.commands.0.id","actual":"t","message":"must not exist"},
"message":"must not exist"},{"attribute":"additionalProperties","property":"devices.0.attributes.1.id","actual":"h","message":"must not exist"},{"attribute":"additionalProperties","property":"devices.0.attributes.0.id","actual":"t","message":"must not exist"},{"attribute":"additionalProperties","property":"devices.0.lazy.0.id","actual":"l","message":"must not exist"}]}
time=2016-02-04T23:26:41.733Z | lvl=DEBUG | corr=n/a | trans=n/a | op=IoTAgentNGSI.NorthboundServer | msg=Error [WRONG_SYNTAX] handing request: Wrong syntax in request: Errors found validating request.
```

**Comment by dmoranj** [comment]
In last versions of the IoTA Library, a small change was introduced in the provisioning API, that may be wrongly specified in some examples. In the specification of attributes:
```
"lazy":[
                  { "id": "l", "name": "luminosity", "type": "percentage" }
            ],
```
the `id` field should be named `object_id`. We have several tasks pending for updating documentation and examples. Next week I'll try to put the documentation up to date. Sorry for the inconvenience.

**Comment by drasko** [comment]
@dmoranj yes, indeed I can notice here: https://github.com/telefonicaid/iotagent-mqtt/blob/master/test/deviceProvisioning/provisionDevice2.json#L19

However, I see that "id" field has disappeared from "attributes" and "commands"... Should we also use "object_id" over there?

What is exact role of "object_id" anyway?

I think this part of doc must be updated: https://github.com/telefonicaid/iotagent-node-lib#-device-provisioning-api

Anyway, this command:
```
(curl localhost:4041/iot/devices -s -S --header 'Content-Type: application/json' \
    --header 'Accept: application/json' --header 'Fiware-Service: mainflux' --header 'Fiware-ServicePath: /mqtt' -d @- | python -mjson.tool) <<EOF
{
    "devices": [ 
        { 
            "device_id": "DevID1", 
            "entity_name": "TheDevice1", 
            "entity_type": "DeviceType", 
            "attributes": [ 
                  { "object_id": "t", "name": "temperature", "type": "float" },
                  { "object_id": "h", "name": "humidity", "type": "float" }
            ],
            "lazy":[
                  { "object_id": "l", "name": "luminosity", "type": "percentage" }
            ],
            "commands": [
                  { "object_id": "t", "name": "turn", "type": "string" }
            ],
            "static_attributes": [
                  { "name": "serialID", "type": "02598347" }
            ]
        }
    ]
} 
EOF
```
passes OK.

My problem is however that I can not see newly added Context Entity in the Orion's MongoDB.
Actually, Orion's log are following:
```
orion_1 | INFO@21:56:44  logMsg.h[1802]: Starting transaction from 172.17.42.1:56184/NGSI9/registerContext
orion_1 | INFO@21:56:44  connectionOperations.cpp[79]: Database Operation Successful (query: { $or: [ { entities: { $in: [ { id: "TheDevice1", type: "DeviceType", isPattern: "false" } ] }, $or: [ { attrs: { $in: [ "luminosity", "turn" ] } }, { attrs: { $size: 0 } } ], expiration: { $gt: 1454795804 } }, { entities.isPattern: "true", expiration: { $gt: 1454795804 }, $where: function(){enId = [ "TheDevice1" ];enType = [ "DeviceType" ];for (var ... } ] })
orion_1 | INFO@21:56:44  connectionOperations.cpp[400]: Database Operation Successful (update: <{ _id: ObjectId('56b66c1c20bef1003a5da142') }, { _id: ObjectId('56b66c1c20bef1003a5da142'), expiration: 1457387804, servicePath: "/mqtt", format: "JSON", contextRegistration: [ { entities: [ { id: "TheDevice1", type: "DeviceType" } ], attrs: [ { name: "luminosity", type: "percentage", isDomain: "false" }, { name: "turn", type: "string", isDomain: "false" } ], providingApplication: "http://localhost:4041" } ] }>)
orion_1 | INFO@21:56:44  logMsg.h[1886]: Transaction ended
orion_1 | INFO@21:56:44  logMsg.h[1802]: Starting transaction from 172.17.42.1:56186/v1/updateContext
orion_1 | INFO@21:56:44  connectionOperations.cpp[79]: Database Operation Successful (query: { _id.id: "TheDevice1", _id.type: "DeviceType", _id.servicePath: /^\/mqtt$/ })
orion_1 | ERROR@21:56:44  safeMongo.cpp[198]: Runtime Error (field 'value' is missing in BSONObj <{ type: "02598347", creDate: 1454795350, modDate: 1454795771 }>)
orion_1 | ERROR@21:56:44  safeMongo.cpp[198]: Runtime Error (field 'value' is missing in BSONObj <{ type: "02598347", creDate: 1454795350, modDate: 1454795771 }>)
orion_1 | ERROR@21:56:44  MongoCommonUpdate.cpp[555]: Runtime Error (unknown attribute value type in DB: 0)
orion_1 | INFO@21:56:44  connectionOperations.cpp[400]: Database Operation Successful (update: <{ _id.id: "TheDevice1", _id.type: "DeviceType", _id.servicePath: /\/mqtt/ }, { $set: { attrs.serialID: { type: "02598347", creDate: 1454795350, modDate: 1454795804 }, modDate: 1454795804 }, $unset: { location: 1 } }>)
orion_1 | INFO@21:56:44  logMsg.h[1886]: Transaction ended
orion_1 | INFO@21:56:49  connectionOperations.cpp[625]: Database Operation Successful (command: { listDatabases: 1 })
orion_1 | INFO@21:56:49  connectionOperations.cpp[79]: Database Operation Successful (query: { conditions.type: "ONCHANGE" })
orion_1 | INFO@21:56:49  connectionOperations.cpp[79]: Database Operation Successful (query: { conditions.type: "ONCHANGE" })
```

**Comment by dmoranj** [comment]
The documentation has to be updated in the IOTA Lib, indeed. BTW, lazy attributes and commands haven't been implemented yet for the MQTT-JSON IoTA. I'll try to create the Step-by-step guide today, to solve your doubts one and for all.

**Comment by drasko** [comment]
@dmoranj no need - it has been done already ;). Take a look at #84, I have already posted you this link: https://github.com/Mainflux/mainflux-fiware/wiki/IoT-Agent-MQTT

I made an extensive step-by-step guide, all you need is to copy-paste it to the IOTA-MQTT page.

**Comment by dmoranj** [comment]
Thanks for the effort :) 

I was thinking of something more structured like the guide we created for the [LWM2M Agent](https://github.com/telefonicaid/lightweightm2m-iotagent/blob/master/docs/userGuide.md#-getting-started). I'm working already on that.

**Comment by dmoranj** [comment]
Closing issue, as it seems to contain already answered questions and some issues regarding the documentation (fixed in #84 or in another PR in the IOTAgent Library documentation).
